### PR TITLE
fix: added support for custom fonts in ensemble studio

### LINKED ISSUE #762: Support custom fonts
**Is your feature request related to a problem? Please describe.**
Fonts are important part of branding and UX. Customers have the ability to upload their own fonts or reference open source fonts like google fonts.

**Describe the solution you'd like**
https://docs.ensembleui.com/theme-and-styling/custom-fonts

**Describe alternatives you've considered**

**Additional context**
EDL implemented in Flutter runtime: https://github.com/EnsembleUI/ensemble/issues/1505

### PR DESCRIPTION
**Pull request by sagardspeed2**
## Describe your changes
Added support for custom fonts in ensemble studio

### Screenshots [Optional]
1. uploaded font
image

2. added classname (`topHeader`) in screen
image

3. added style (`fontFamily` in `topHeader` class) to class in theme
image

4. output
image

## Issue ticket number and link

Closes #762 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have added tests
- [x] I have added a changeset `pnpm changeset add`
- [x] I have added example usage in the kitchen sink app

### CODE REVIEW

File: packages/framework/src/loader.ts [hunk 1]
```diff
@@ -70,8 +73,20 @@ const getArtifacts = async (
         ...config,
         secretVariables: document.secrets as { [key: string]: unknown },
       };
+    } else if (document.type === "font") {
+      const font = document as { [key: string]: unknown };
```
Comment (evshi) [inline packages/framework/src/loader.ts:77]: FontDTO type should reflect what the firebase document is like and should be what you cast to here.

File: packages/framework/src/hooks/useFonts.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { useEffect } from "react";
+import { error, type EnsembleFontModel } from "../shared";
+
+export const useFonts = (fonts: EnsembleFontModel[]): void => {
+  useEffect(() => {
+    const fontFaces = fonts.map((font) => {
+      return new FontFace(font.family, `url(${font.url})`, {});
```
Comment (evshi) [inline packages/framework/src/hooks/useFonts.ts:7]: What about the font options?

**Comment by sagardspeed2** [comment]
@evshi this PR is still not ready for review, I am still working on it

**Comment by amin-nas** [comment]
@sagardspeed2 @evshi please see latest EDL example here:
https://github.com/EnsembleUI/ensemble/issues/1576

File: packages/framework/tsconfig.json [hunk 3]
```diff
@@ -1,5 +1,8 @@
 {
   "extends": "tsconfig/react-library.json",
   "include": ["./src"],
-  "exclude": ["dist", "build", "node_modules"]
+  "exclude": ["dist", "build", "node_modules"],
+  "compilerOptions": {
+    "lib": ["dom", "dom.iterable"]
+  }
```
Comment (evshi) [inline packages/framework/tsconfig.json:7]: this should go in the base config.

File: packages/framework/src/shared/dto.ts [hunk 4]
```diff
@@ -55,3 +56,19 @@ export interface EnsembleConfigYAML {
...
 }
+
+export interface FirebaseFontDTO {
+  readonly fontFamily: string;
+  readonly publicUrl: string;
+  readonly fontWeight: string;
+  readonly fontStyle: string;
+}
+
+export interface FontDTO {
```
Comment (evshi) [inline packages/framework/src/shared/dto.ts:67]: Just keep `EnsembleFontModel` and remove this, rename the FirebaseFontDTO to FontDTO. Convert things later so models reference models and dtos reference dtos, etc.

DTO = Data Transfer Object, it represents the data over the wire.

File: packages/framework/src/hooks/useFonts.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        await fontFace
+          .load()
+          .then((loadedFont) => {
+            document.fonts.add(loadedFont);
+          })
+          .catch((e) => {
+            error(e);
+          });
+      });
+    };
```
Comment (evshi) [inline packages/framework/src/hooks/useFonts.ts:21]: You can just use `forEach` and you don't need a new function declaration ehre.

File: packages/framework/src/hooks/useApplicationContext.tsx [hunk 6]
```diff
@@ -36,6 +37,9 @@ export const ApplicationContextProvider: React.FC<
     );
   });
 
+  // load custom fonts
+  useFonts(app.fonts || []);
```
Comment (evshi) [inline packages/framework/src/hooks/useApplicationContext.tsx:41]: Handle nullability in the hook itself.

**Review by evshi — CHANGES_REQUESTED** [review]

File: packages/framework/src/hooks/useFonts.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { useEffect } from "react";
+import { error, type EnsembleFontModel } from "../shared";
+
+export const useFonts = (fonts?: EnsembleFontModel[]): void => {
+  const loadFontFace = (font: EnsembleFontModel): unknown => {
```
Comment (evshi) [inline packages/framework/src/hooks/useFonts.ts:5]: nit: just inline this or put it inside the hook body so it's not redeclared.

**Review by evshi — APPROVED** [review]

**Comment by evshi** [comment]
@sagardspeed2 is this ready to be merged?

**Comment by sagardspeed2** [comment]
> @sagardspeed2 is this ready to be merged?

@evshi we need to run this in production first to clear the cors issue
`gsutil cors set cors.json <bucket_name>`

and I don't have the production firebase access

**Comment by evshi** [comment]
@sagardspeed2 why do we need to run that command? @justEhmadSaeed is this something you can help with?

**Comment by sagardspeed2** [comment]
@evshi we are trying to fetch the font files from the firebase and there is cors added for the origin. so when I tested it was giving cors error while fetching the font files from the firebase. so I used this command for dev firebase storage, and it works.

**Comment by justEhmadSaeed** [comment]
@sagardspeed2 i don't understand why you need to set cors for fonts when it's already being fetched by the prod app. if you want to fetch the prod data on the local, then it's disabled intentionally.

**Pull request merged**